design-docs/improved-console.md

```diff
@@ -0,0 +1,283 @@
+ - a fixed-height, random-access `TextArea` (progressArea)
+ 
+The mainArea is displayed at the top for rendering important messages like warnings
+and errors.
+
+`--quiet`, `--info`, and `--debug` log level flags affect the output visible in the main TextArea,
+However, the default will change so that `LIFECYCLE` logs are no longer displayed.
+ 
+The `ProgressLabel` is intended to give the user a glance-able indicator that tells
+them whether the build will be finished soon or not. 
+See ["Display build progress as a progress bar through colorized output"](TODO) below.
+
+The multiple progressArea lines are rendered results of a fixed list of `AbstractWorker`s. 
+It displays work-in-progress for each `AbstractWorker`. 
+Forked processes submit BuildOperations with information about "what" is running "where".
+
+Plugins can provide progress indication by publishing `BuildOperations`
+
+A `OperationStatusBarFormatter` can be used to customize the intra-line display of each 
+`ProgressOperation` chain (each `ProgressOperation` has a reference to its parent)
+
+Implementation is similar to `ConsoleBackedProgressRenderer` with throttling and a data
+structure for storing recent updates.
+
+> NOTE: We must avoid use of save/restore cursor position unless we switch to a 
+[terminfo](https://en.wikipedia.org/wiki/Terminfo)-based Console handling instead of using ANSI.
+    
+### Test Coverage
+* A terminal size with fewer than max parallel operations shows (rows - 2) operations in parallel
+* Operation status lines are trimmed at (cols - 1)
+* `System.in` and SystemConsole I/O happens on the mainArea, which is unaffected by other areas
+* Renders well on default macOS Terminal, default Windows Console, Cygwin, and common Linux Terminals
+* Console display degrades gracefully given BuildOperations that are out-of-order (e.g. complete event received before start)
+
+### Open issues
+* How is `--continuous` build displayed? During execution? While waiting? (as we allow processes to stay running)
+* When things fail, will the new CLI look like a regular failure or something else?
+* How does `--parallel` (or not) affect display?
+* How does having more operations than workers affect display?
+
+## Story: Display build progress as a progress bar through colorized output
+Render ProgressEvents with build completeness information as a progress bar.
+ 
+Intended to give the user a very fast way of telling whether the build will be finished soon or not.
+The work-in-progress (yellow) section shows how much of the build would be complete if all
+Operations displayed in the progressArea completed.
+
+### User-visible Changes
+We can visually represent complete, in-progress, and un-started tasks of build using colorized output:
+
+`#####green#####>##yellow##>#####black#####> 40% Building`
+
+On ANSI terminals, we can use empty spaces with different background colors for the "bars" and
+A red background shall be used for failed tasks. 
+
+### Implementation
```

Comment:
I hope there aren't any changes to TAPI necessary, just handling of additional types of BuildOperations. That has yet to be speced